src/charm.py

```diff
@@ -99,7 +99,15 @@ def __init__(self, *args):
             source_type="loki",
             source_url=self._external_url,
         )
-        scrape_jobs = [{"static_configs": [{"targets": [f"*:{self._port}"]}]}]
```

Comment:
This is star notation because we need it to automatically expand into all units.
This goes into app data so only the leader does that. The current change would only include the leader unit. I.e. currently the proposed change is buggy.

I.e. you must use star notation.

Add utest as well?

Reply:
This is also related to line 119
```
port=parsed_external_url.port or self._port
```
which is also buggy. It should probably be
```
port=parsed_external_url.port or 80
```

Reply:
Other than the fact Loki does not, in fact, listen on `80` by default. It's more of a concern that `urlparse.ParseResult` doesn't have a `port` field at all, and that IPU returns a "bare" URL which doesn't include a port anyway.

Probably, it will be necessary to either make `_external_url` a little more complex to actually figure out what's happening instead of just blindly returning `ingress_url`, to _assume_ that if `ingress_url` is valid (maybe an attribute on the class, or changing the type of `_external_url`, which would be more invasive) the port is 80, or other.

Reply:
Fixed.